Fix avhrr gac geometry

oembury opened the pull request:
Updates the avhrr_gac instrument definition to match the across track resolution of AVHRR GAC data.

Previously the definition appears to copy+paste from the full resolution avhrr instrument definition, and did not account for the reduced across track resolution (409 pixels instead of 2048)


 - [x] Closes #184
 - [ ] Tests added <!-- for all bug fixes or enhancements -->
 - [ ] Fully documented <!-- remove if this change should not be visible to users, e.g., if it is an internal clean-up, or if this is part of a larger project that will be documented later -->

oembury commented:
Also updates the avhrr instrument definition to accept either "number of scans" or "array of scan times" for consistency with avhrr_gac

mraspaud reviewed (changes requested):
Thanks an lot for your contributions! Really nice that you took the time also to add documentation and tests.
It looks good in general, but I'd like to have the new functionality moved to new functions to avoid introducing breaking changes. See the inline comments for details.

oembury commented on pyorbital/geoloc_instrument_definitions.py:
```diff
@@ -81,22 +92,34 @@ def avhrr_gac(scan_times, scan_points,
     Source: NOAA KLM User's Guide, Appendix J
     http://www.ncdc.noaa.gov/oa/pod-guide/ncdc/docs/klm/html/j/app-j.htm
+
+    :scan_times: Number of scanlines (int) or array of Observation times (datetime object)
```
This one - avhrr_gac accepting either a number of lines OR list of datetimes - was existing behaviour

oembury commented:
I've reverted the changes and added two new functions avhrr_from_times / avhrr_gac_from_times as requested.

These will only accept time delta / datetime like inputs, and will not accept a "number of scanlines" as done by avhrr_gac.

However, they will accept a numeric input (interpreted as number of seconds "time delta" relative to start of scanning) for consistency with `ScanGeometry`.

mraspaud commented on pyorbital/geoloc_instrument_definitions.py:
```diff
@@ -100,6 +100,99 @@ def avhrr_gac(scan_times, scan_points,
+    raise TypeError(f'unsupported type for ScanGeometry times: {type(times)}')
+
+
+def avhrr_from_times(scan_times, scan_points, scan_angle=55.37):
+    """Definition of the avhrr instrument.
+
+    Source: NOAA KLM User's Guide, Appendix J
+    http://www.ncdc.noaa.gov/oa/pod-guide/ncdc/docs/klm/html/j/app-j.htm
+
+    :scan_times: Observation times or offsets
+    :scan_points: Across track pixel positions
+    :scan_angle: Maximum scan angle of the outermost FOV
+    """
+
+    offset = _calc_time_offsets(scan_times)
+    scan_points = np.asanyarray(scan_points)
+    scans_nb = len(offset)
+
+    avhrr_inst = np.vstack([(scan_points / 1023.5 - 1) * np.deg2rad(-scan_angle),
+                            np.zeros(len(scan_points))])
+
+    avhrr_inst = np.tile(avhrr_inst[:, np.newaxis, :], [1, scans_nb, 1])
+    # building the corresponding times array
+    times = (np.tile(scan_points * 0.000025, [scans_nb, 1])
+             + np.expand_dims(offset, 1))
+    return ScanGeometry(avhrr_inst, times)
```
I think having the first argument being either a time or a delta is ambiguous and can lead much more headscratching than would be reasonable in the future. Why not split the functionality? for example:

```python
def avhrr_from_times(scan_times, scan_points, scan_angle=55.37):
    """Definition of the avhrr instrument given times.

    Source: NOAA KLM User's Guide, Appendix J
    http://www.ncdc.noaa.gov/oa/pod-guide/ncdc/docs/klm/html/j/app-j.htm

    :scan_times: Observation times
    :scan_points: Across track pixel positions
    :scan_angle: Maximum scan angle of the outermost FOV
    """

    offsets = _calc_time_offsets(scan_times)
    retura avhrr_from_time_offsets(offsets, scan_points, scan_angle)
    
def avhrr_from_time_offsets(offsets, scan_points, scan_angle)
    """Definition of the avhrr instrument given time offsets.

    Source: NOAA KLM User's Guide, Appendix J
    http://www.ncdc.noaa.gov/oa/pod-guide/ncdc/docs/klm/html/j/app-j.htm

    :scan_times: Time offsets from reference time
    :scan_points: Across track pixel positions
    :scan_angle: Maximum scan angle of the outermost FOV
    """
    scan_points = np.asanyarray(scan_points)
    scans_nb = len(offset)

    avhrr_inst = np.vstack([(scan_points / 1023.5 - 1) * np.deg2rad(-scan_angle),
                            np.zeros(len(scan_points))])

    avhrr_inst = np.tile(avhrr_inst[:, np.newaxis, :], [1, scans_nb, 1])
    # building the corresponding times array
    times = (np.tile(scan_points * 0.000025, [scans_nb, 1])
             + np.expand_dims(offsets, 1))
    return ScanGeometry(avhrr_inst, times)
```

oembury replied:
I think having two functions could be more confusing to an end-user.

If there had been two functions last week I might have assumed that `avhrr_from_times` and `avhrr_from_time_offsets` were doing something different. i.e. `avhrr_from_times` must be storing the actual scanline times so I would not have to pass the scanline times to compute_pixels etc as I've already stored them in the avhrr instrument object.

How about changing the doc string to:
```
    :scan_times: Along track scan time as offset from start of scanning.
        May be passed as timedelta or number of seconds. If datetime objects
        are supplied then offset is calculated as (scan_times - scan_times[0])
```

mraspaud reviewed (commented):
Thanks for making new functions! I think it's looking good, but a few things could be simplified/refactored

oembury commented on pyorbital/geoloc_instrument_definitions.py:
```diff
@@ -100,6 +100,99 @@ def avhrr_gac(scan_times, scan_points,
+    :scan_angle: Maximum scan angle of the outermost FOV
+    """
+
+    offset = _calc_time_offsets(scan_times)
+    scan_points = np.asanyarray(scan_points)
+    scans_nb = len(offset)
+
+    # The AVHRR swath is +/- 55.4 degrees, which puts the outermost FOV at
+    #   55.4 * 1023.5 / 1024 = 55.373
+    # GAC pixels are the average of four FOVs with sampling: ..1111.2222.----.NNNN..
+    # so we need to reduce the scan_angle by a factor (1023.5 - 3.5) / 1023.5
+    # to calculate the GAC pixel centres
+    gac_angle = scan_angle * (1023.5 - 3.5) / 1023.5
+
+    avhrr_inst = np.vstack([(scan_points / 204 - 1) * np.deg2rad(-gac_angle),
+                            np.zeros(len(scan_points))])
+
+    avhrr_inst = np.tile(
+        avhrr_inst[:, np.newaxis, :], [1, np.int32(scans_nb), 1])
+    # building the corresponding times array
+    times = (np.tile(scan_points * 0.000125, [scans_nb, 1])
+             + np.expand_dims(offset, 1))
+    return ScanGeometry(avhrr_inst, times)
```
I'd prefer to leave that change until after resolving the issue about where the GAC tie-points are actually located (i.e. pytroll/pygac#88).

I think your proposed wrapper is correct (and very nearly wrote it that way in the first place!) but didn't want to introduce another transformation of the x coordinate right now...

mraspaud replied:
sure, you’re right, best to write tests first, refactor afterwards

oembury commented on pyorbital/geoloc_instrument_definitions.py:
```diff
@@ -100,6 +100,99 @@ def avhrr_gac(scan_times, scan_points,
+    :times: Array-like of times or offsets
+    """
+
+    try:
+        # Convert timedelta to total seconds
+        return np.array([t.total_seconds() for t in times])
+    except (TypeError, AttributeError):
+        pass
```
There is no need, but I added it to help me understand what input pyorbital was asking for compared to what it actually needed.  It seemed to me that there are a few layers of functions that convert times back and forth between different forms. From pygac it appears to be:

* [pygac] Determine missing lines as np.timedelta64
* [pygac] Add scanline[0] time to get np.datetime64
* [pygac] convert to datetime.datetime
* [pyorbital] subtract first missing line to get datetime.timedelta
* [pyorbital] convert to float seconds
* [pyorbital] convert to np.timedelta64

When calculating the positions it will:

* [pygac] get first missing line (np.datetime64), convert to datetime.datetime
* [pyorbital] convert to np.datetime64 and add to stored np.timedelta64

As all the instrument definitions are basically building a ScanGeometry with scanline times stored as np.timedelta64 offsets from the start of the scan it seemed that the primary "time" input to a `*_from_times` function should be np.timedelta64 - or something that has an obvious duck typing to np.timedelta64

Ideally a future version of `ScanGeometry` should also be modified so it would also accept "times" in the format of np.timedelta64 rather than only as (float) seconds - reducing the number of conversions required and avoiding the current rounding error

mraspaud replied:
Ok, it sounds indeed like we are moving back and forth between types from what you show, and that’s obviously not optimal.
It seems to me that the cause of these is because the interfaces in pyorbital were build before numpy’s time types were a thing, but it make sense nowadays to migrate to these. So I would say maybe the first step is to make the `ScanGeometry` accept numpy datetime64.
When it comes to providing timedelta(64), I can see that it is tempting because internally timedeltas are used in both pygac and pyorbital, but I would argue that it is an implementation detail and is confusing to the user: the logical interface for such a function imo is to accept times, not timedeltas, as the user wants the positions of pixel for given times. Also, since the number of scanlines is limited to a few thousands at most, I would say the performance hit to work with datetime64 at the interface level is minimal.
So in conclusion, I would say:
- make `ScanGeometry` accept times as datetime64
- have `avhrr_from_times` accept scan times as datetime(64)

Do you agree?

oembury replied:
That would be fine for my use case: calculating AVHRR GAC geolocation. However, I'd be concerned about the impact on any other use of pyorbital so I'd prefer not to attempt it in this PR..

Are you proposing that `ScanGeometry` should accept np.datetime64 instead of float seconds. Or would it accept both?
* All the other instrument definitions just take a number of scanlines and calculate the delta times (as float seconds) to pass to `ScanGeometry`. If the interface was changed to only np.datetime64 then you'd have to add a nominal start date to all geometries.
* Removing support for float seconds would break the API for anyone using `ScanGeometry` directly
* Supporting both np.datetime64 and "time deltas as float seconds", but not np.timedelta64 is confusing (I guess you would mark the "time deltas as float seconds" with a deprecation warning for removal?)

Overall I agree that `ScanGeometry` and the instrument definition should accept times in the same format. Or have a common "as_times" function that could be used by any both to ensure consistency. I'm not familiar enough with pyorbital to say what the users will want.

Anyway - this is probably best as a separate issue / PR. For now should I just remove the code from this PR and go back to only accepting objects which behave like datetime.datetime?

mraspaud replied:
> For now should I just remove the code from this PR and go back to only accepting objects which behave like datetime.datetime?

Yes that sound reasonable.

mraspaud commented on pyorbital/geoloc_instrument_definitions.py:
```diff
@@ -100,6 +103,63 @@ def avhrr_gac(scan_times, scan_points,
+    Source: NOAA KLM User's Guide, Appendix J
+    http://www.ncdc.noaa.gov/oa/pod-guide/ncdc/docs/klm/html/j/app-j.htm
+
+    :scan_times: Observation times or offsets
```
Just times, not offsets?

mraspaud reviewed (approved):
Apart from two typos (I think?), I’m good with this :+1:
